tsinfer/inference.py

```
@@ -1444,85 +1443,116 @@ def __init__(self, sample_data, ancestor_data, time_units=None, **kwargs):
-            # Create a list of all ID ranges in each epoch.
-            breaks = np.where(self.epoch[1:] != self.epoch[:-1])[0]
-            start = np.hstack([[0], breaks + 1])
-            end = np.hstack([breaks + 1, [self.num_ancestors]])
-            self.epoch_slices = np.vstack([start, end]).T
-            self.num_epochs = self.epoch_slices.shape[0]
-        self.start_epoch = 1
-
-    def __epoch_info_dict(self, epoch_index):
-        start, end = self.epoch_slices[epoch_index]
-        return collections.OrderedDict(
-            [("epoch", str(self.epoch[start])), ("nanc", str(end - start))]
+        for t in self.ancestor_data.ancestors_time[:]:
+            self.tree_sequence_builder.add_node(t)
+
+    def group_by_linesweep(self):
+        start = self.ancestor_data.ancestors_start[:]
+        end = self.ancestor_data.ancestors_end[:]
+        time = self.ancestor_data.ancestors_time[:]
+
+        # We only need to perform the grouping for the small epochs at earlier times.
+        # Skipping the later epochs _really_ helps as later ancestors are dependent on
+        # almost all the earlier ones, so the dependency graph becomes intractable.
+        breaks = np.where(time[1:] != time[:-1])[0]
+        epoch_start = np.hstack([[0], breaks + 1])
+        epoch_end = np.hstack([breaks + 1, [self.num_ancestors]])
+        time_slices = np.vstack([epoch_start, epoch_end]).T
+        epoch_sizes = time_slices[:, 1] - time_slices[:, 0]
+        # To choose a cutoff point find the first epoch that is 50 times larger than
+        # the median epoch size. For a large set of human genomes the median epoch
+        # size is around 10, so we'll stop grouping by linesweep at 5000.
+        try:
+            first_large_epoch = np.where(epoch_sizes > 500 * np.median(epoch_sizes))[0][
+                0
+            ]
+            first_large_epoch_first_ancestor = time_slices[first_large_epoch, 0]
+        except IndexError:
```

Not clear why we get an IndexError here. I think it's better to have a conditional on the actual cause here rather than letting exceptions handle things, as it can mask genuine errors.

I guess it's something like
```python
cutoff =  500 * np.median(epoch_sizes)
if np.max(epoch_sizes) < cutoff:
    first_large_epoch = np.where(..)
else:
   # other thing
```

---

Done